src/server/journal/types.h

```diff
@@ -44,6 +48,35 @@ struct Entry {
   uint64_t expire_ms = 0;  // 0 means no expiry.
 };
 
+// This struct represents a single journal entry.
+// Those are either control instructions or commands.
+struct EntryNew {  // Called this "New" because I can't delete the old neither replace it partially
+  // Payload represents a non-owning view into a command executed on the shard.
+  using Payload =
+      std::variant<std::monostate,                        // No payload.
+                   CmdArgList,                            // Parts of a full command.
+                   std::pair<std::string_view, ArgSlice>  // Command and its shard parts.
+                   >;
+
+  // Owned version of Payload.
+  using OwnedPayload = std::optional<CmdArgVec>;
+
+  EntryNew(TxId txid, DbIndex dbid, Payload pl)
+      : txid{txid}, opcode{Op::VAL}, dbid{dbid}, payload{pl}, owned_payload{} {
+  }
+
+  EntryNew(journal::Op opcode, DbIndex dbid)
+      : txid{0}, opcode{opcode}, dbid{dbid}, payload{}, owned_payload{} {
+  }
+
+  TxId txid;
+  Op opcode;
+  DbIndex dbid;
+
+  Payload payload;
+  OwnedPayload owned_payload;
```

Comment:
payload is used for writing journal entries and owned_payload used for reading entries. I would consider using 2 struct one for writer and one for reader.

Reply:
Why? What does it simplify or make more coherent?

We can make it extend a common base and use two subtypes for the varying subfields

Reply:
It simplifies understanding code and flow. You wrote the code so you understand that on the master and in writer you need to use payload and in replica and reader you need to use owned_payload, but for someone who reads it or want to change or add things to the flow I need to read the whole flow to understand it.
You could add a comment to the struct on this but it will be better to enforce it buy using 2 types and than when I add to the flow I cant access the incorrect filed.
Of course you can use base class.